### PR TITLE
qp-gen: strengthen reliability, improve latency

### PR DESCRIPTION
**Pull request by mamuqsit**
## Introduction
Question generation relied on synchronous OpenAI calls with loosely enforced JSON formatting, requiring manual parsing and post-processing. This often led to inconsistent structures, missing fields such as `keyAnswer`, and fragile handling of different question types (especially for matching-type questions).

## Approach
Generation is fully migrated to the async API. QP gen flow now utilizes Structured Outputs to conform to defined Pydantic models at API boundary. System now returns strongly typed objects instead of raw dictionaries, meaning the need for manual JSON cleaning is eliminated.

Schema handling has been simplified to avoid embedding full JSON schemas in prompts. Question models are enriched with clearer field semantics (particularly around `keyAnswer` and match-the-following structures).

## Changes
- tweak: use vendor-provided async apis over blocking calls to qdrant, openai chat completion https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/a6457097d83209ebbfe8dad2636dc06423d74006, https://github.com/A4i-tech/Shiksha-Copilot/pull/70/changes/916f0fe8c265d11f82e33de7de707bb5aab693ba
- bug fix: ai-generated questions often failed due to malformed json outputs from the llm. new approach uses openai's responses api for structured outputs and internal retries, effectively offloading the entire modeling and parsing cost to openai https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/aa9bbf7634d3231cd9783dfa1497e1e26a49d2e1
- bug fix: mcqs often comprised of 0, 3, or 5 choices despite being intended to contain strictly 4 choices https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/5877d0b6753e6e3dcfd881b1d87a3312a1588411
- bug fix: ai-generated question tasks often failed due to `No question found for Normalized key: four alternatives are given for each of the following questions, choose the correct alternative|1|prose: the noble bishop|knowledge` - this was the cause of llm not generating a correct value for `unit_name` field because of confusing prompting. for instance, the prompt would read `The unit_name should reference a valid chapter/unit from the noble bishop` instead of `The unit_name should reference a valid chapter/unit from **Chapters** above.` https://github.com/A4i-tech/Shiksha-Copilot/pull/70/changes/77b38d93eb867c2fd185ee90999341223afc6b31
- bug fix: qb-cache never expired. not only did it serve outdated ai-generated question schemas, this also caused ai-generation to often fail entirely due to schema errors https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/1d03f47b3fad05aeee1c8ca70e130929abd97dbb
- bug fix: questions were stripped of several fields before being stored in cache, causing questions retrieved from cache to have undefined difficulty, marks, and even an empty options set (in case of mcqs) https://github.com/A4i-tech/Shiksha-Copilot/pull/70/changes/afb771ee25929eaf74c187ad577cca78f84eaed6
- cleanup: remove unnecessary /question-bank-cache rest apis https://github.com/A4i-tech/Shiksha-Copilot/pull/70/changes/e57c9a43c78aa2a07741740fd13f5b10c13a64bc
- cleanup: remove unused, unneeded embedding lookup mechanism in qp-cache https://github.com/A4i-tech/Shiksha-Copilot/pull/70/changes/2f9159a8c75adc380fc283705f71f2dd37a185a9
- cleanup: reduce excessive use of try..catch in shiksha-api https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/5f05e071d592794b71aca79f5f88d59ec6bbc36b
- cleanup: remove unnecessary shiksha-api response validators from shiksha-backend https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/23b5e08c7c2baefb8a97e3b10b34ccf51405b7f9

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
Below is your complete, rules‑compliant code review.

---

# High‑Level Summary

This PR is extremely large. Most changes are positive (async OpenAI usage, RAG rewrite, removal of embeddings infra, strict typing, cache TTL, cleaned router).  

However, the PR also introduces several **high‑severity logic bugs**, **schema mismatches**, and **error‑handling violations** in hot‑path services (question generation, caching, and translation).  

The most severe issues are in:

- QuestionPaperService (structured output logic)
- Cache queue manager (dangerous field mutations)
- FastAPI router error handling regressions
- OpenAI config typing regression

These will reliably break production generation or create silently corrupt cache entries.

---

# Issues (Prioritized)

## CRITICAL

### 1. Wrong type for `azure_openai_deployment_name`
**File:** `app/config.py`, class `Settings`  
**Issue:** Field type changed from `str` to `ResponsesModel`. Environment variables load as strings → this will crash settings initialisation or cause runtime serialization errors.  
**Fix:** Restore to `Optional[str]`, store model name as string.

---

### 2. RAG adapter returns incorrect type breaking generation
**File:** `rag_adapter_cache.py`, method `chat_with_index`  
**Issue:** Switched from `chat.completions.create` to `client.responses.create` but still return `response.output_text`. That text may not be JSON, yet caller expects JSON and passes to `model_validate_json()`.  
**Fix:** Ensure RAG adapter always returns valid JSON by wrapping user prompt with explicit schema requirement and raising if content is not valid JSON.

---

### 3. Broken JSON parsing logic for structured output
**File:** `question_paper_service.py`, `_generate_questions_batch`  
**Issue:** For non-RAG path, switched to `responses.parse` (good), but RAG path still tries to manually strip `'json` wrappers and parse raw strings. Upstream LLM may return markdown fenced code → parsing will crash or silently drop output.  
**Fix:** Unify both paths: always require RAG adapter to produce valid `GeneratedQuestionItemResponse` JSON, or route everything through OpenAI structured responses.

---

### 4. Cache insertion logic mutates live Mongoose docs
**File:** `cache.queue.manager.js`, `buildCacheAdditions`  
**Issue:** It spreads a Mongoose doc into a plain object, mutates nested arrays, and later sends into `updateCache`, implicitly overwriting fields. High risk of removing fields, losing indexes, corrupting documents.  
**Fix:** Explicitly clone only whitelisted fields: `{chapterId, unitName, unitLevel, questions, questionsToAdd}`.

---

### 5. Missing error logging for translation failures violates ERROR HANDLING rule
**File:** `question_paper.py`, `translate_json_content_to_kannada`  
**Issue:** When translation fails, only raises HTTPException, missing required structured logging of function name & input summary.  
**Fix:** Add `logger.error(..., extra={ "function": "...", "input": ... })` before raising.

---

## HIGH

### 6. `GeneratedQuestionItem.item` excludes MCQ option nesting
**File:** `question_paper.py`, model `GeneratedQuestionItem`  
**Issue:** `item: Union[...]` requires full model, but upstream generator sometimes returns flat dicts (e.g. in tests). `model_validate` will throw.  
**Fix:** Add pre-validator to coerce dicts into appropriate DTO before model validation.

---

### 7. `get_question_distribution` now ignores template verification failures
**File:** `question_paper_service.py`, method `get_question_distribution`  
**Issue:** After the structured parse, the verification error is logged but not raised → invalid templates propagate.  
**Fix:** Raise HTTP 400 if verification fails.

---

### 8. Changed translation API signature breaks Node backend call sites
**File:** `question_bank_manager.js`, `_handleTranslation`  
**Issue:** Manager now calls backend as `(target_language, json_data)` but translation endpoint expects `target_language` and `json_data` from Body. Some Node callers still send `{target_language, json_data}`. This will break.  
**Fix:** Normalize interface: always POST `{"target_language": "...", "json_data": {...}}`.

---

## MEDIUM

### 9. No timeout for async OpenAI and Qdrant requests
**Files:**  
`question_paper_service.py`, all OpenAI calls  
`rag_adapter_cache.py`, all Qdrant calls  
**Issue:** Violates API rule: external calls must set explicit timeouts → risk of blocking generation.  
**Fix:** Add `timeout=15` or use httpx client with configured global timeout.

---

### 10. Inconsistent difficulty propagation
**File:** `question_paper_service.py`, `_organize_questions_into_response`  
**Issue:** Difficulty is no longer injected into MCQ/Text items when already wrapped inside `GeneratedQuestionItem`. Some consumers expect item.difficulty.  
**Fix:** In casting step: ensure final question objects include difficulty.

---

### 11. Cache TTL applied on `updatedAt` may wipe in-progress jobs
**File:** `question.bank.cache.model.js`  
**Issue:** TTL uses updatedAt → any worker not updating documents within TTL loses data mid-flight.  
**Fix:** Use dedicated TTL field (e.g., `expiresAt`) or set TTL only on “stale version”.

---

## LOW

### 12. Logging prints raw LLM prompts (PII/content leakage)
**Files:** `question_paper_service.py`, `_generate_questions_batch`  
**Issue:** Logs system prompt and user message verbatim → violates LOGGING rule about sensitive content.  
**Fix:** Mask content or remove logs.

### 13. Inconsistent casing for fields (`unit_name` vs `unitName`)
Will cause cache misses and mismatches.  
**Fix:** Normalize before writing or reading.

---

# Score Calculation

CRITICAL: 5 × -5 = -25  
HIGH: 3 × -3 = -9  
MEDIUM: 3 × -2 = -6  
LOW: 2 × -1 = -2  

Total deductions = 42  
Score = 100 - 42 = **58**

Sanity check: Code still functions but with correctness and cache risks → score in high 50s is appropriate.

---

# Optional References

- OpenAI Structured Responses: https://platform.openai.com/docs/guides/structured-outputs  
- FastAPI Logging Best Practices: https://fastapi.tiangolo.com/advanced/custom-logging/  
- MongoDB TTL Collections Guidance: https://www.mongodb.com/docs/manual/core/index-ttl/

---

If you want, I can generate **patch-ready fixes** for all CRITICAL/HIGH issues.

**Review by a4i-architect — COMMENTED** [review]
Below is your complete code review.

---

# ✅ High‑Level Summary

This PR is large and improves async usage, structured outputs, cache correctness, and removes dead code. However, several newly introduced issues create **runtime failures**, **broken types**, **timeout risks**, and **API contract regressions**.  
The most severe issues are:

- Incorrect type for `azure_openai_deployment_name`
- Missing logging of full context for errors (required by global standards)
- Missing timeouts on external API calls
- Possible RAG adapter failure due to incorrect parameter shape
- Type mismatches causing model validation failures in generated Pydantic models
- Removed unit tests for removed features broke coverage guarantees

---

# 🚨 Issues (Prioritized)

## CRITICAL ISSUES (‑5 each)

### 1. Wrong type for `azure_openai_deployment_name`
**File:** `app/config.py`  
**Issue:** Modified field type to `ResponsesModel` which is **not a string**, but the service expects a **deployment name string**. This will break SDK calls and configuration loading.  
**Fix:** Change type back to `Optional[str]`.

---

### 2. Missing timeouts on every external call  
**Files:**  
- `question_paper_service.py` (multiple OpenAI calls)  
- `rag_adapter_cache.py` (Qdrant + OpenAI embeddings)  

**Issue:** Violates API standard: *“All external API calls must have explicit timeout”*.  
Lack of timeouts = **hanging worker threads & stuck event loop**.  
**Fix:** Pass `timeout=...` to all SDK calls (`client.responses.parse`, `client.responses.create`, `client.embeddings.create`, `qdrant.search`).  

---

### 3. Incorrect use of `input=` in RAG adapter  
**File:** `rag_adapter_cache.py`, function `chat_with_index`  
**Issue:** `AsyncOpenAI.responses.create` does **not** accept a list of messages as `input`. This breaks RAG completely.  
**Fix:** Pass a single string or restructure to use `responses.parse`.

---

### 4. GeneratedQuestionItem misuse  
**File:** `question_paper_service.py`, `_generate_questions_batch`  
**Issue:** When parsing RAG output, `GeneratedQuestionItemResponse.model_validate_json` returns `items` as **Pydantic models**, but `_organize_questions_into_response` expects `GeneratedQuestionItem` fields like `.type`.  
If the generated JSON deviates, this will raise exceptions.  
**Fix:** Validate `.output_parsed` strictly and fallback on errors, logging full context.

---

## HIGH ISSUES (‑3 each)

### 5. No structured logging for Translation endpoint  
**File:** `routers/question_paper.py`  
**Issue:** Uses `logger.info(str)` but lacks context (function + inputs) violating logging standards.  
**Fix:** Attach dict-based context: `logger.info("...", extra={"target_language": target_language})`.

---

### 6. RAG Adapter not closing Qdrant connection on failures  
**File:** `rag_adapter_cache.py`, `cleanup()`  
**Issue:** If initialization errors occur, the connection may remain open.  
**Fix:** Wrap in try-finally for closing.

---

### 7. New cache TTL index uses `updatedAt` but updates do not always mutate the timestamp  
**File:** `question.bank.cache.model.js`  
**Issue:** Pushing new questions updates the doc, but some flows write only into nested projection; may not bump TTL. Leads to premature TTL eviction.  
**Fix:** Ensure writes always update `updatedAt` or set `{ timestamps: true }`.

---

### 8. Missing tests for new service-level functions  
**Files:**  
- `cache.queue.manager.js` (new logic)  
- New `buildCacheAdditions` logic  

**Issue:** Violates *Testing Standard*: “New service-level functions require an explicit unit test”.  
**Fix:** Add test coverage for both functions.

---

## MEDIUM ISSUES (‑2 each)

### 9. Schema instructions changed but format rules still reference removed fields  
**File:** `_get_format_instruction_for_type`, removal of `.schema_dict()`  
**Issue:** Old instructions tell LLM to follow non-existent JSON schema props → weaker model guidance.  
**Fix:** Provide explicit static JSON schema or switch to `text_format=` only.

---

### 10. `_format_system_prompt` no longer catches errors  
**File:** `question_paper_service.py`  
**Issue:** Removed try/except. Any formatting error now blows up generation.  
**Fix:** Reintroduce try-except with context logging.

---

### 11. Missing `await` on `client.responses.create` error branch  
**File:** `rag_adapter_cache.py`  
**Issue:** Response-handling inconsistency may lead to unhandled promises.  
**Fix:** Ensure every SDK call is awaited.

---

### 12. Potential duplication of questions due to key normalization mismatch  
**File:** `_organize_questions_into_response`  
**Issue:** Normalization logic mismatches `_build_generation_slots`. Causes missing lookup → fallback warning → empty questions.  
**Fix:** Use central normalize function everywhere.

---

## LOW ISSUES (‑1 each)

### 13. Missing type annotations in modified functions  
Affected files include:  
- `cache.queue.manager.js`  
- `_buildCacheAdditions`

Reduces maintainability.

---

### 14. Unused imports remain  
**File:** `routers/question_paper.py` → `Detector` imported but rarely used.  
**Fix:** Remove unused imports.

---

### 15. Repeated JSON dumps in prompts harms performance  
**File:** `_format_system_prompt`  
Minor but unnecessary overhead.

---

# 🧮 Score Calculation

| Severity | Count | Deduction |
|---------|--------|-----------|
| CRITICAL | 4 | −20 |
| HIGH | 4 | −12 |
| MEDIUM | 4 | −8 |
| LOW | 3 | −3 |
| **Total Deduction** |  | **−43** |

**Final Score = 100 − 43 = 57**

Sanity check: All CRITICAL/HIGH issues relate to real breakages (wrong types, broken RAG, async timeouts missing). Score 57 is reasonable.

---

# 📚 Optional References

- OpenAI Responses API: https://platform.openai.com/docs/api-reference/responses
- Qdrant Async client: https://qdrant.tech/documentation/clients/python/
- Python logging structure patterns: https://docs.python.org/3/howto/logging-cookbook.html

---

If you want, I can generate **inline patch fixes** for each CRITICAL issue.

**Review by a4i-architect — COMMENTED** [review]
Below is your complete code review following all rules.

---

# High‑Level Summary

This PR is extremely large and does major refactors: removing embedding paths, migrating to async OpenAI APIs, restructuring cache, tightening Pydantic models, and reworking question‑paper generation. Most changes are improvements.

However, there are **several CRITICAL and HIGH‑severity issues** mainly around:

- Silent error swallowing in async LLM/RAG paths  
- Structural narrowing of schemas breaking backwards compatibility  
- Missing error handling for external API failures violating team standards  
- Incorrect logging of sensitive data  
- Missing awaits and incorrect assumptions about async client contracts  
- Cache update paths that can corrupt data or drop questions

The actionable list below focuses solely on added/modified lines.

---

# Issues & Fixes

## CRITICAL

### 1. Missing timeout for *all* external API calls  
**File:** `question_paper_service.py`, functions `_generate_questions_batch`, `get_question_distribution`, `generate_question_bank_by_parts`  
**Issue:** Calls to `AsyncAzureOpenAI.responses.parse` / `.create` specify no explicit timeout. Violates API standard #1 → risk of hanging worker.  
**Fix:** Pass `timeout=settings.openai_timeout_seconds` to every OpenAI call.

---

### 2. Sensitive data logged  
**File:** `question_paper.py`, endpoint `/by-parts`  
**Issue:** `logger.info` logs full generated payload including questions, possibly containing user content → violates Logging rule #3.  
**Fix:** Replace with masked logs or log only metadata fields.

---

### 3. Silent exception swallowing corrupts question generation  
**File:** `question_paper_service.py::_generate_questions_batch`  
**Issue:** Broad `except Exception` returns `[]` which discards all generated questions without surfacing error → question paper silently incomplete.  
**Fix:** Log error at ERROR level with context and re‑raise or wrap in ValueError so upstream returns 500.

---

### 4. Cache TTL Index will delete documents unexpectedly  
**File:** `question.bank.cache.model.js`  
**Issue:** Added TTL index expiring every document after 6h unconditionally. This removes user‑generated cache used for correctness.  
**Fix:** TTL must **only** apply to stale versions (e.g., add `expireAfterSeconds` on a separate timestamp or add flag). Otherwise generation becomes unstable.

---

### 5. Broken fallback unit matching may cause mismatched or misrouted questions  
**File:** `question_paper_service.py::_organize_questions_into_response`  
**Issue:** If LLM returns unknown `unit_name`, code silently falls back to normalized original name, causing mismatched grouping and wrong marking distribution.  
**Fix:** Instead of fallback, **raise** ValueError when LLM returns unknown unit. Provide safe retry.

---

## HIGH

### 6. Breaking change to MCQ options schema  
**File:** `question_paper.py > FourOptionsQuestion`  
**Issue:** Now requires exactly 4 options via `min_length=4, max_length=4`, breaking backward compatibility with existing data and UI paths expecting variable lengths.  
**Fix:** Only enforce strict length on *generation*, not on model validation (e.g., use custom validator).

---

### 7. Missing error handling for langdetect raising exceptions  
**File:** `question_paper.py::translate_json_content_to_kannada`  
**Issue:** `.detect()` may throw `LangDetectException`, but this code removed try/except from previous version.  
**Fix:** Wrap detection in try/except and log at WARNING level as before.

---

### 8. RAG adapter cleanup called unconditionally in error path  
**File:** `question_paper_service.py::_get_or_create_rag_adapter`  
**Issue:** `await adapter.cleanup()` is called even when adapter failed initialization and may be partially constructed → risk of None access.  
**Fix:** Check `if adapter and hasattr(adapter, 'cleanup')`.

---

### 9. `GeneratedQuestionItemResponse.model_validate_json` may throw unhandled JSON errors  
**File:** `question_paper_service.py::_generate_questions_batch`  
**Issue:** RAG output is free‑form text; passing raw content into strict parser without try/except will crash.  
**Fix:** Wrap parse in try/except and surface 500.

---

### 10. Unsafe logging of user-supplied JSON in `/translate-json`  
**File:** `translate_json_content_to_kannada`  
**Issue:** Logs entire JSON user payload at INFO level.  
**Fix:** Log only metadata (sizes, key counts) not raw content.

---

## MEDIUM

### 11. Missing structured logging everywhere added  
**File:** `lesson_chat_service.py`, `question_paper.py`  
**Fix:** Replace logger.info/f"…" with `logger.info("message", extra={…})`.

---

### 12. DAO aggregation risks `$or` array empty case → full collection scan  
**File:** `question.bank.cache.dao.js`  
**Fix:** Protect with `if questionFilters.length > 0` else skip `$or`.

---

### 13. Missing `await` for `.save()` in new doc creation  
**File:** `cache.queue.manager.js::updateCache`  
**Fix:** Ensure all save calls are `await newDoc.save()`.

---

### 14. LLM instructions in `_generate_questions_batch` inconsistent with structured outputs  
**File:** `question_paper_service.py`  
**Issue:** Using `"Ensure item adheres to json_schema"` while using OpenAI structured output model → instructions redundant/confusing.  
**Fix:** Remove these manual schema hints.

---

### 15. Potential race between TTL deletion and cache update job  
**File:** `cache.queue.manager.js`  
**Fix:** Add transaction or version check.

---

## LOW

### 16. Missing type annotations in many added helper functions  
**File:** `cache.queue.manager.js`  
**Fix:** Add JSDoc for `buildCacheAdditions` and `_buildCacheQuestionFilters`.

---

### 17. Removed test coverage of some cases (e.g., schema_dict)  
**File:** various test deletions  
**Fix:** Add replacement tests for structured output flow.

---

# Score Calculation

CRITICAL: 5 × -5 = -25  
HIGH: 5 × -3 = -15  
MEDIUM: 4 × -2 = -8  
LOW: 2 × -1 = -2  

Total deductions = 25 + 15 + 8 + 2 = 50

**Final Score = 100 – 50 = 50**

---

# References (non‑obvious API behaviors)

- OpenAI Python SDK timeout behavior  
  https://platform.openai.com/docs/guides/error-codes/python-timeouts

- FastAPI error handling best practices  
  https://fastapi.tiangolo.com/tutorial/handling-errors/

---

If you want, I can produce a patch diff fixing all CRITICAL and HIGH issues.

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The PR is large and generally improves async usage, removes legacy embedding logic, and aligns responses with new OpenAI structured outputs. However, several correctness and reliability issues exist, mostly around error handling, missing fallbacks, incomplete cache‑update logic, and broken RAG adapter API usage.

Below are only issues on lines *added or modified* in this PR.

---

## ❗ Critical Issues

### 1. **RAG adapter: misuse of the OpenAI Responses API**  
**File:** `rag_adapter_cache.py`, function `chat_with_index`  
**Issue:** The code now calls `client.responses.create` with a `messages` array inside `input`, which is not valid for the OpenAI 2.x Responses API. Responses API expects a text/string, not chat message objects. This will break all RAG calls.  
**Fix:** Convert messages → a single concatenated string, or move to `client.chat.completions.create` with correct models.  
**Severity:** CRITICAL (-5)

### 2. **Missing await on qdrant.close() leads to leaked connections**  
**File:** `rag_adapter_cache.py`, function `cleanup`  
**Issue:** `await self.qdrant.close()` is correct, but other call sites (e.g., error paths) do not await cleanup, causing resource leaks.  
**Fix:** Ensure every call site that invokes `cleanup()` awaits it.  
**Severity:** CRITICAL (-5)

### 3. **Structured Output: failure paths drop all generation**  
**File:** `question_paper_service.py`, `_generate_questions_batch`  
**Issue:** The code returns `[]` on *any* exception during structured output parse, silently discarding content, causing downstream missing questions.  
**Fix:** Log error with context, rethrow as ValueError so caller can fail with 500 instead of returning partial paper.  
**Severity:** CRITICAL (-5)

---

## ❗ High Severity Issues

### 4. **Incorrect normalisation of question keys produces systematic cache misses**  
**File:** `question_paper_service.py`, `_organize_questions_into_response`  
**Issue:** `item` is now a typed Pydantic model, but code still assumes dict-like structures when building `question_directory`, losing difficulty and failing type matching.  
**Fix:** Use `generated.item` directly and ensure `cast()` is applied in only one layer.  
**Severity:** HIGH (-3)

### 5. **Translation endpoint: no timeout on third-party call**  
**File:** `question_paper.py`, endpoint `/translate-json`  
**Issue:** Violates global API rule: external API call must have explicit timeout.  
**Fix:** Add timeout parameter in `TranslationService.translate_json_async`.  
**Severity:** HIGH (-3)

### 6. **Cache update logic: inconsistent schema (questions vs. questionsByObjective)**  
**File:** `question_bank.cache.helper.js`, modified sections  
**Issue:** Cache model changed from `questionsByObjective` → flat `questions[]`, but many functions still rely on objective-based nesting when matching filters.  
**Fix:** Update matching, merging, and pool logic to operate on new flat schema consistently.  
**Severity:** HIGH (-3)

### 7. **buildCacheAdditions incorrectly derives keys**  
**File:** `cache.queue.manager.js`  
**Issue:** Uses raw Pydantic objects vs. plain dicts inconsistently; unit name lowercasing is missing in some downstream comparisons.  
**Fix:** Normalise unit name and objective consistently at creation time.  
**Severity:** HIGH (-3)

---

## ⚠️ Medium Severity Issues

### 8. **_get_format_instruction_for_type forms invalid instructions**  
**File:** `question_paper_service.py`  
**Issue:** It prints `{qtype}` object string rather than model name/schema → LLM may not follow format.  
**Fix:** Use `qtype.model_name()` consistently.  
**Severity:** MEDIUM (-2)

### 9. **Missing structured logging for translation source language detection failures**  
**File:** `question_paper.py`  
**Issue:** Logs raw sample text, could leak user content (violates logging policy: no sensitive content).  
**Fix:** Mask content.  
**Severity:** MEDIUM (-2)

### 10. **RAG cleanup missing in some fallback paths**  
**File:** `question_paper_service.py`, `_get_or_create_rag_adapter`  
**Issue:** After catching initialization errors, the code calls cleanup but does not guard adapter existence.  
**Fix:** Check `if adapter:` before cleanup.  
**Severity:** MEDIUM (-2)

### 11. **No unit tests for new cache-addition logic**  
**File:** `cache.queue.manager.js`  
**Issue:** Substantial logic added (pooling, limits, counting) but no corresponding unit tests. Violates testing rules.  
**Fix:** Add tests for: limit enforcement, duplicate avoidance, objective matching.  
**Severity:** MEDIUM (-2)

---

## 🔎 Low Severity Issues

### 12. **Missing type hints for many newly added functions**  
**File:** `cache.queue.manager.js`, `question_paper_service.py`  
**Severity:** LOW (-1)

### 13. **Direct use of random() in shouldUseCache**  
**File:** `question.bank.cache.helper.js`  
**Issue:** Unstable behaviour across runs, makes debugging hard.  
**Fix:** Inject seed or make deterministic for tests.  
**Severity:** LOW (-1)

### 14. **Inconsistent naming: `marks_per_question` vs `marksPerQuestion`**  
**File:** Multiple  
**Issue:** Could break filters and templates.  
**Severity:** LOW (-1)

---

## Final Score

Start: 100  
Critical: 3 × -5 = -15  
High: 4 × -3 = -12  
Medium: 4 × -2 = -8  
Low: 3 × -1 = -3  

**Final Score = 100 - 38 = 62**

No reclassification needed.

---

## References

- OpenAI Responses API official spec: https://platform.openai.com/docs/guides/responses  
- Pydantic model parsing best practices: https://docs.pydantic.dev/latest/  
- FastAPI timeout patterns: https://fastapi.tiangolo.com/advanced/custom-request-timeouts/

**Comment by mamuqsit** [comment]
@a4i-architect 

1. Responses API misuse
OpenAI’s Responses API `input` accepts both - string and message objects with `role` and `content`. So `input=messages` is valid. See docs: https://platform.openai.com/docs/api-reference/responses/compact?api-mode=responses.

2. Missing await on qdrant.close()
`cleanup()` awaits `self.qdrant.close()`. Cache eviction and full cache cleanup paths also await adapter cleanup. I don’t see an unawaited production call site.

3. Typed Pydantic model handled like dict
`_organize_questions_into_response()` uses attributes: `generated.type`, `generated.unit_name`, `generated.item` - it is not treating `GeneratedQuestionItem` as a dict there. Also the template matching key does not include difficulty, but the template itself does not appear to request difficulty, so "losing difficulty" is not a bug.

4. Translation endpoint no timeout
Endpoint does not pass timeout but Azure translator client is constructed with `connection_timeout=10` and `read_timeout=60`.

5. Cache schema inconsistent: questions vs questionsByObjective
Cache helper reads flat `cacheDoc.questions || []` and matches flat model.

File: shiksha-api/app-service/app/services/question_paper_service.py [hunk 1]
```diff
@@ -293,54 +301,45 @@ def _format_system_prompt(
...
+            GRADE=request.grade,
+            SUBJECT=request.subject,
+            TOTAL_MARKS=request.total_marks,
+            CHAPTERS=unit_name,  # Single unit for this batch
+            UNIT_WISE_LEARNING_OUTCOMES=unit_los_text,
+            EXISTING_QUESTIONS_JSON=json.dumps(
+                existing_questions, ensure_ascii=False
+            ),
+            QUESTION_BANK_BLOOM_TAXONOMY_GUIDE=blooms_guide,
+            GRAMMAR_TOPICS=self._get_grammar_topics(request),
```
Comment (N1kunj1998) [inline shiksha-api/app-service/app/services/question_paper_service.py:337]: Since we have removed GRAMMAR_TOPICS from prompt this will be silently dropped either we should add prompt back or remove this line

File: shiksha-api/app-service/app/models/question_paper.py [hunk 2]
```diff
@@ -29,7 +29,12 @@ class QuestionBankMetadata(BaseModel):
 class TextQuestion(BaseModel):
     question: str = Field(default="")
     answer: str = Field(default="")
-    keyAnswer: str = Field(default="")
+    keyAnswer: str = Field(default="", description="\n".join([
+        "Answer to be displayed right below the question."
```
Comment (N1kunj1998) [inline shiksha-api/app-service/app/models/question_paper.py]: missing comma at the end of line 33

File: shiksha-api/app-service/app/routers/question_paper.py [hunk 3]
```diff
@@ -127,468 +54,101 @@ def get_sample_text(data: Any) -> str:
...
-            "description": "Returns the (potentially) translated JSON object.",
-            "model": TranslationResponse,
-        },
-        status.HTTP_500_INTERNAL_SERVER_ERROR: {
-            "description": "An unexpected error occurred while processing the JSON."
-        },
-    },
-    tags=["Utilities"] 
-)
+@router.post("/translate-json", summary="Translate JSON Content (Auto-Detect Source)")
```
Comment (N1kunj1998) [inline shiksha-api/app-service/app/routers/question_paper.py:57]: Route endpoint name is changed here is this intentionally done?, `translate_json` → `translate-json`

File: shiksha-api/app-service/app/services/question_paper_service.py [hunk 4]
```diff
@@ -34,15 +36,26 @@
...
+
+
+class GeneratedQuestionItemResponse(BaseModel):
+    items: list[GeneratedQuestionItem]
+
+
 class QuestionPaperService:
     """Service for handling question paper generation using Azure OpenAI."""
 
+    chat_deployment: ResponsesModel
```
Comment (N1kunj1998) [inline shiksha-api/app-service/app/services/question_paper_service.py:50]: in line 59 we are assigning chat_deployment a string value while here we are using class level type annotation which is wrong type, we should change this to string.

**Review by N1kunj1998 — COMMENTED** [review]
Good Implementation Just few small bugs and regressions that we need to take care of.

File: shiksha-api/app-service/app/services/question_paper_service.py (hunk 1 above)
Comment (mamuqsit) [reply shiksha-api/app-service/app/services/question_paper_service.py:337]: GRAMMAR_TOPICS existed in 2 places in the prompt, so I removed the duplicate occurrence - there's one `{GRAMMAR_TOPICS}` where this replacement would happen.

https://github.com/A4i-tech/Shiksha-Copilot/blob/b8f14d90d5221907c58020a872567290c5f0e4af/shiksha-api/app-service/prompts/question_paper_prompts.yaml#L8-L20

File: shiksha-api/app-service/app/models/question_paper.py (hunk 2 above)
Comment (mamuqsit) [reply shiksha-api/app-service/app/models/question_paper.py]: Addressed in https://github.com/A4i-tech/Shiksha-Copilot/pull/70/commits/0ecda518dd0e891212162c7856dddd2c7aaaf38e

image

File: shiksha-api/app-service/app/routers/question_paper.py (hunk 3 above)
Comment (mamuqsit) [reply shiksha-api/app-service/app/routers/question_paper.py:57]: Yeah this was intentional - there was a route named /by-parts, I made the change so we consistently stick to either kebab-case or under_scores.

No standards prohibit using underscores but using kebab-case is the norm for url paths:
- microsoft azure api guidelines say [DO use kebab-casing (preferred) or camel-casing for URL path segments](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md)
- google's aip-122 on resource naming uses kebab-case in examples: https://google.aip.dev/122

File: shiksha-api/app-service/app/services/question_paper_service.py (hunk 4 above)
Comment (mamuqsit) [reply shiksha-api/app-service/app/services/question_paper_service.py:50]: `ResponsesModel` is a type-alias for str - figured it would be useful to do early-validation for incoming model names, or as a developer hint for now.

image

**Review by N1kunj1998 — APPROVED** [review]
LGTM

**Review by theainerd — APPROVED** [review]
LGTM. This is a strong reliability improvement.

- Structured Outputs, async clients, and TTL-based caching directly address the four reported bug fixes.
- Moving from a worker thread to an in-process PQueue also makes sense, since cache updates are now lightweight and no longer require embedding hops.
